include/bouncer.h

```diff
@@ -32,6 +32,9 @@
 #include <event2/event.h>
 #include <event2/event_struct.h>
 
+#include "uthash.h"
```

Comment:
@turnstep I'm moving the discussion on hash functions here to not clutter the main issue thread. For reference this was your original message:

> FWIW, I did some experimenting with the various hash functions (see https://troydhanson.github.io/uthash/userguide.html#hash_functions) using various contrived scenarios to represent "real world" traffic. 
> 
> tl;dr JEN is good, and so is BER
> 
> For most scenarios, things are pretty equal, with no clear winner. Even runs with the some large differences still had JEN doing well, if we value the add/find times more than the ideal distribution (which I think we should given the small hash size) The SQL queries I used were all 500 characters long, against Postgres 15.This is the keystats run for a test with 200 different queries with a 100 queue, and then executing a random one 10,000 times:
> 
> ```
> fcn  ideal%     #items   #buckets  dup%  fl   add_usec  find_usec  del-all usec
> ---  ------ ---------- ---------- -----  -- ---------- ----------  ------------
> FNV   95.4%        606        128  845%  ok       1580       1643            50
> BER   93.6%        606        128  845%  ok        488        501            21
> SFH   88.8%        606        256  845%  ok        847        809           127
> JEN   87.0%        606        256  845%  ok        348        340            15
> SAX   86.5%        606        128  845%  ok       1125       1503            48
> OAT   86.3%        606        128  845%  ok        986       1010            18
> ```
> 
> The worst I got JEN was with a queue of 100, 1000 random queries, and then picking 200 at random 10,000 times:
> 
> ```
> fcn  ideal%     #items   #buckets  dup%  fl   add_usec  find_usec  del-all usec
> ---  ------ ---------- ---------- -----  -- ---------- ----------  ------------
> BER   97.4%        606        128  855%  ok        498        479            23
> FNV   97.0%        606        128  855%  ok        902        918            27
> OAT   92.6%        606        128  855%  ok        666        699            13
> SFH   91.7%        606        128  855%  ok        574        603            32
> SAX   89.6%        606        256  855%  ok        542        556            27
> JEN   85.5%        606        128  855%  ok        981        960            33
> ```
> 
> This scenario would argue for BER, which is much better here and holds up pretty well in all the other scenarios as well. One more scenario: queue size of 100, but prepare 50 and pick from those 50 randomly:
> 
> ```
> fcn  ideal%     #items   #buckets  dup%  fl   add_usec  find_usec  del-all usec
> ---  ------ ---------- ---------- -----  -- ---------- ----------  ------------
> BER  100.0%        156         32    0%  ok         77         81             3
> SFH  100.0%        156         32    0%  ok         77         85             4
> FNV  100.0%        156         32    0%  ok        135        141             5
> JEN  100.0%        156         32    0%  ok        120        127             5
> SAX  100.0%        156         32    0%  ok        110        117             5
> OAT   90.4%        156         64    0%  ok        245        246             5
> ```
> 
> (Aside: how is OAT so bad at this? 😀 )
> 
> 
> Obviously, figuring out what a typical real-world scenario would be is the tricky part! I suspect ORMs in particular probably create more unique queries than they ought to, which would mean a lot of churn in the queue. 
>

Reply:
Could you share the commands/code/datafiles you used to get these numbers?

One thing that I'm gathering from your description is that you tried hashing SQL queries. But that is actually not the most common hashing operation in this implementation. Hashing a SQL query only happens when a `Parse` packet is sent, because that's when the client hash entry needs to be linked to the global cache. The most common hashing operations are:
1. hashing the statement name, to find the matching entry in the client hash. 
2. hashing the internal queryid, to check if the statement is prepared on the server

Those two happen on every `Bind` packet, so every time a prepared statement is executed. So I think we should optimize the hash function for those two cases, instead of the hashing of the query (`Parse` is expected to add some tracking overhead).

Reply:
I did a combination of packets. Just some super simple things like this:

```
my $LONGSQL = "SELECT count(*), 12345 FROM pg_class where relpages < ?
AND relname NOT IN ('alphabet1','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2',
'alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2',
'alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2','alphabet2',
'alphabet2','alphabet2','alphabet2','alphabet2','alphabet2345')";

## Scenario 8: prepare 50 different queries, execute random ones 10,000 times
if (8 == $scenario) {

  for my $x (1..50) {
      $SQL = $LONGSQL;
      $SQL =~ s/12345/12345."$x"/;
      $s{$x} = $dbh->prepare($SQL, { pg_server_prepare=>1 });
  }
  for (1 .. 10000) { my $x = int rand(50); $x++; $s{$x}->execute(10); }
}
```

Again, not sure what actual real-world things look like. Personally, the things I have written would be even a smaller pool than the above - maybe 10 or less statements, all called millions of times. So yes, agree we should focus on `Bind`

Reply:
I did some more tests for all the types of hashmaps that we used and indeed BER seemed best. So I now started using that one.